### PR TITLE
Haskell updates

### PR DESCRIPTION
**Pull request by domenkozar**

### CODE REVIEW

**Comment by domenkozar** [comment]
@GrahamcOfBorg build cachix niv cabal2nix

**Comment by domenkozar** [comment]
@GrahamcOfBorg eval

**Review by peti — CHANGES_REQUESTED** [review]
:-1: 

This branch is in no shape for a merge. Some of the changes to `configuration-hackage2nix.yaml` didn't even have a chance to take effect yet because the generator wasn't run since they were added.

**Comment by domenkozar** [comment]
Sure, I'd like to understand what's missing and help out. I should have opened a draft PR, but too late :)

**Comment by domenkozar** [comment]
@GrahamcOfBorg build cachix niv cabal2nix

**Comment by domenkozar** [comment]
@peti https://hydra.nixos.org/eval/1545587 looks good

File: pkgs/development/haskell-modules/configuration-nix.nix [hunk 1]
```diff
@@ -94,11 +94,9 @@ self: super: builtins.intersectAttrs super {
   # Won't find it's header files without help.
   sfml-audio = appendConfigureFlag super.sfml-audio "--extra-include-dirs=${pkgs.openal}/include/AL";
 
-  cachix = overrideCabal (addBuildTools (enableSeparateBinOutput super.cachix) [pkgs.boost]) (drv: {
-    postPatch = (drv.postPatch or "") + ''
-      substituteInPlace cachix.cabal --replace "c++14" "c++17"
-    '';
-  });
+  cachix = addBuildTools (enableSeparateBinOutput super.cachix) [pkgs.boost];
```
Comment (peti) [inline pkgs/development/haskell-modules/configuration-nix.nix]: What part of `boost` does `cachix` need? Could you please document that in a comment?

**Review by peti — CHANGES_REQUESTED** [review]

File: pkgs/development/haskell-modules/configuration-nix.nix (hunk 1 above)
Comment (domenkozar) [reply pkgs/development/haskell-modules/configuration-nix.nix]: See 821676200d6f65c9eb4584e393af1204eeb7e363

**Comment by domenkozar** [comment]
@peti what's left?

**Comment by peti** [comment]
> @peti what's left?

Haddock is currently broken.

**Review by peti — APPROVED** [review]

**Review by peti — CHANGES_REQUESTED** [review]
We need a working haddock, I suppose: https://hydra.nixos.org/eval/1546016

**Comment by domenkozar** [comment]
OK. I'm going to make another release of cachix and we can merge this tomorrow.

**Comment by domenkozar** [comment]
Released, if you update the set we can fix haddock and merge.

**Comment by domenkozar** [comment]
```
diff --git a/pkgs/development/haskell-modules/configuration-hackage2nix.yaml b/pkgs/development/haskell-modules/configuration-hackage2nix.yaml
index 70dddd52252..4604a941ae8 100644
--- a/pkgs/development/haskell-modules/configuration-hackage2nix.yaml
+++ b/pkgs/development/haskell-modules/configuration-hackage2nix.yaml
@@ -2488,11 +2488,7 @@ extra-packages:
   - gloss < 1.9.3                       # new versions don't compile with GHC 7.8.x
   - haddock < 2.17                      # required on GHC 7.10.x
   - haddock == 2.17.*                   # required on GHC 8.0.x
-  - haddock-api == 2.15.*               # required on GHC 7.8.x
-  - haddock-api == 2.16.*               # required on GHC 7.10.x
-  - haddock-api == 2.17.*               # required on GHC 8.0.x
-  - haddock-library == 1.2.*            # required for haddock-api-2.16.x
-  - haddock-library == 1.4.3            # required for haddock-api-2.17.x
+  - haddock-api == 2.22.*               # 2.23.* breaks on GHC 8.6
   - happy <1.19.6                       # newer versions break Agda
   - haskell-gi-overloading == 0.0       # gi-* packages use this dependency to disable overloading support
   - haskell-src-exts == 1.19.*          # required by hindent and structured-haskell-mode
```

This should fix it I guess.

**Comment by domenkozar** [comment]
@peti does the above patch look good?

**Comment by peti** [comment]
> @peti does the above patch look good?

Yes, it looks good to me. I'm not sure whether other haddock packages will build properly with a version of `haddock-api` that's older than their own ... but they might. I haven't checked.

**Comment by domenkozar** [comment]
@peti cool, if you can regenerate the packages I'll verify they build and we can merge.

**Comment by peti** [comment]
> Ff you can regenerate the packages I'll verify they build and we can merge.

Done. In case you are not aware of the video, https://www.youtube.com/watch?v=qX0mgtSm360 might be of interest.

**Comment by domenkozar** [comment]
```
$ nix-build -A haskellPackages.haddock
/nix/store/whfsdjijy0arma080lw8s087gl11pbsy-haddock-2.22.0
```

It seems to work again. Going to merge after CI passes the evaluation checks.

**Comment by peti** [comment]
I would not have merged the branch in this state. I would have liked to do more clean-up first. :disappointed:

**Comment by domenkozar** [comment]
I really wanted nixpkgs master fixed, seems like the cleanup can be done on top of this work? :)

**Comment by peti** [comment]
Well, if you don't want to take my interests seriously, then I guess I'll just file that under "experience".

**Comment by domenkozar** [comment]
@peti I've asked you (multiple times) in this PR:

- that I'd really like to get this merged
- what do you think is missing

You approved the PR, then said we have to fix haddock. I've shown you the diff I applied (which I simplified).

So I'm quite certain I did my best to give you a chance to communicate what you'd like to do.

**Comment by domenkozar** [comment]
That said, thanks for the video, I'll watch it.

**Pull request merged**